daemon/emer-daemon.c

```diff
@@ -1299,12 +1320,60 @@ on_timer_stopped_cb (EmerAggregateTimer     *timer,
   else
     emer_aggregate_timer_complete_stop_timer (timer, invocation);
 
+  /* Stop watching the sender if this was the last of its timers */
+  sender_name = emer_aggregate_timer_impl_get_sender_name (timer_impl);
+  sender_data = g_hash_table_lookup (priv->monitored_senders, sender_name);
+  g_assert (sender_data != NULL);
+  g_ptr_array_remove_fast (sender_data->aggregate_timers, timer_impl);
+  if (sender_data->aggregate_timers->len == 0)
+    g_hash_table_remove (priv->monitored_senders, sender_name);
+
   if (!g_hash_table_remove (priv->aggregate_timers, timer_impl))
     g_warning ("Stopped timer was not in the set of running timers");
 
   return TRUE;
 }
 
+static void
+sender_name_vanished_cb (GDBusConnection *connection,
+                         const gchar     *sender_name,
+                         gpointer         user_data)
+{
+  EmerDaemon *self = EMER_DAEMON (user_data);
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+  AggregateTimerSenderData *sender_data;
+  g_autoptr(GDateTime) now = NULL;
+  gint64 now_monotonic_us;
+  guint i;
+
+  now = g_date_time_new_now_local ();
+  now_monotonic_us = g_get_monotonic_time ();
+
+  sender_data = g_hash_table_lookup (priv->monitored_senders, sender_name);
+  g_assert (sender_data != NULL);
+
+  for (i = 0; i < sender_data->aggregate_timers->len; i++)
+    {
+      g_autoptr(GError) error = NULL;
+      EmerAggregateTimerImpl *timer_impl =
+        g_ptr_array_index (sender_data->aggregate_timers, i);
+
+      emer_aggregate_timer_impl_stop (timer_impl,
+                                      now,
+                                      now_monotonic_us,
+                                      &error);
+
+      if (error)
+        g_warning ("Error stopping aggregate timer after sender disappeared: %s",
+                   error->message);
+
+      if (!g_hash_table_remove (priv->aggregate_timers, timer_impl))
+        g_warning ("Stopped timer was not in the set of running timers");
+    }
+
+  g_hash_table_remove (priv->aggregate_timers, sender_name);
```

Comment:
This is the wrong hash table. `aggregate_timers` is a map-as-set whose keys are EmerAggregateTimerImpl.

I am surprised this doesn't crash when calling emer_aggregate_timer_impl_hash on a string!

Before fixing this bug, I suggest adjusting `emer_aggregate_timer_impl_hash` and `emer_aggregate_timer_impl_equal` to use checked casts of their arguments:

```diff
diff --git a/daemon/emer-aggregate-timer-impl.c b/daemon/emer-aggregate-timer-impl.c
index 4294a09..5ebeede 100644
--- a/daemon/emer-aggregate-timer-impl.c
+++ b/daemon/emer-aggregate-timer-impl.c
@@ -192,8 +192,7 @@ emer_aggregate_timer_impl_stop (EmerAggregateTimerImpl  *self,
 guint
 emer_aggregate_timer_impl_hash (gconstpointer timer_impl)
 {
-  const EmerAggregateTimerImpl *self =
-    (EmerAggregateTimerImpl *)timer_impl;
+  const EmerAggregateTimerImpl *self = EMER_AGGREGATE_TIMER_IMPL (timer_impl);
 
   return g_str_hash (self->cache_key_string);
 }
@@ -202,8 +201,8 @@ gboolean
 emer_aggregate_timer_impl_equal (gconstpointer a,
                                  gconstpointer b)
 {
-  const EmerAggregateTimerImpl *timer_impl_a = (EmerAggregateTimerImpl *)a;
-  const EmerAggregateTimerImpl *timer_impl_b = (EmerAggregateTimerImpl *)b;
+  const EmerAggregateTimerImpl *timer_impl_a = EMER_AGGREGATE_TIMER_IMPL (a);
+  const EmerAggregateTimerImpl *timer_impl_b = EMER_AGGREGATE_TIMER_IMPL (b);
 
   return g_str_equal (timer_impl_a->cache_key_string,
                       timer_impl_b->cache_key_string);
```

Then testing the code should log a critical.